CRM/Core/Permission/Joomla.php

```diff
@@ -131,4 +135,74 @@ public function checkGroupRole($array) {
     return FALSE;
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function upgradePermissions($permissions) {
+    $translatedPerms = array();
+    foreach (array_flip($permissions) as $perm) {
+      $translated = $this->translateJoomlaPermission($perm);
+      $translatedPerms[] = $translated[0];
+    }
+
+    $associations = $this->getUserGroupPermsAssociations();
+    $originalAssociations = clone $associations;
+    foreach (array_keys(get_object_vars($associations)) as $permName) {
+      if (!in_array($permName, $translatedPerms)) {
+        unset($associations->$permName);
+      }
+    }
+
+    if ($originalAssociations !== $associations) {
+      $this->updateGroupPermsAssociations($associations);
+    }
+  }
+
+  /**
+   * Fetches the associations between user groups and CiviCRM permissions.
+   *
+   * @return object
+   *   Properties of the object are Joomla-fied permission names.
+   */
+  private function getUserGroupPermsAssociations() {
+    // Get a db connection.
```

Comment:
Since no one in the CiviCRM space seems to know anything about Joomla! internals, why not be verbose with comments?